design/ingressroute-design.md

```diff
@@ -0,0 +1,398 @@
+This property makes the design proposed appealing from a usability standpoint as _most_ ingress usecases are simple--publish my web app on this url--so it feels right to favour a design that does not penalise the default, simple, use case.
+
+- links to other proposals
+
+# Future work
+
+Future work outside the scope of this design includes:
+
+- Limit the set of namespaces where root IngressRoutes are valid. Only those permitted to operate in those namespaces can therefore create virtual hosts and delegate the permission to operate on them to other namespaces. This would most likely be acomplished with a command line flag or ConfigMap.
+- Delegation to matching labels, rather than names, may be added in the future. This is valid, as long as none of the matching IngressRoute objects are roots, because routes are a set, so can be merged from several objects _in the same namespace_.
+
+# Security Concerns
+
+As it relates to the CRD the security implications of this proposal, the model of delegating a part of a vhost's route space to a CRD in another namespace is implies a level of trust.
+I, the owner of the virtual host, delegate to you the /foo prefix and everything under it, implies that you cannot operate on this vhost outside /foo, but it also means that whatever you do in /foo; host malware, leak k8s session keys, is something I'm trusting you not to do.
+
+## No ability to prevent route delegation
+
+In writing up this proposal an issue occured to me.
+It happens when Contour is operating in the, poorly specified, enforcing mode; where Contour only recognises root IngressRoutes in a set of authorised namespaces.
+(well, it happens regurdless of enforcement mode, but the argument is that if you do not turn on enforcement you are explicitly saying you trust you users, or you have another mechanism in place -- CI/CD -- to handle this).
+
+Imaging this scenario:
+```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  # this is www.google.com's root object
+  name: google
+  namespace: virtualhost-namespace
+spec:
+  virtualhost:
+    fqdn: www.google.com
+    tls:
+      # required, the name of a secret in the current namespace
+      secretName: google-tls
+      # other properties like cipher suites may be added later
+  routes:
+  - match: /mail
+    # delegate to the gmail service running in another namespace
+    delegate:
+      name: gmail
+      namespace: gmail
+    ...
+```
+And in the gmail namespace we have a IngressRoute vertex
+```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: gmail 
+  namespace: gmail
+spec:
+  routes:
+  - match: /mail
+    service:
+      name: gmail-tomcat-svc
```

Comment:
Haha -- gmail is java but not tomcat.

Reply:
/me mumbles something about artistic liberties.